plugins/out_s3/s3.c

```diff
@@ -101,6 +115,44 @@ struct flb_aws_header *create_canned_acl_header(char *canned_acl)
     return acl_header;
 };
 
+static struct flb_aws_header *create_headers(struct flb_s3 *ctx, int num_headers)
+{
```

Comment:
In any language, it's good to design each function to do a specific job. As much as possible, you want to make sure it only does a single job, and encapsulates everything needed for that job. And then give it the fewest number of arguments possible

In this case- I think this function can calculate `num_headers` itself using the `ctx`. That encapsulates all behavior related to creating headers in this function. 

Outside you need to use the num_headers of course, so this function needs to return that information.

In C there is a very common pattern for doing this:

```
static int create_headers(struct flb_s3 *ctx, struct flb_aws_header **headers, int num_headers)
```

Basically, to return multiple pieces of information, at multiple pointer arguments. The return value is 0 or -1 depending on whether the function was successful or encountered an error.

You then call it like this:
```
struct flb_aws_header *headers;
int num_headers;

ret = create_headers(ctx, &headers, &num_headers);
```

And then in the function, allocate the headers as you do now, and then "return" them at the end like this:

```
*headers = s3_headers;
*num_headers = header_len;
```

Something like that.